File: VBterminal/terminal/Form1.vb
' terminal emulator - in Visual Studio VB

Public Class Form1
    ' set up delegate to display data received from serial port
    Private Delegate Sub SerialDelegate(ByVal Buffer As String)
    Private adre As New SerialDelegate(AddressOf DisplayData)

    ' key pressed on terminal textbox
    Private Sub TextBox1_KeyPress(ByVal sender As System.Object, ByVal e As System.Windows.Forms.KeyPressEventArgs) Handles TextBox1.KeyPress
        Console.WriteLine("key hit" + e.KeyChar)
        If EchoOFFToolStripMenuItem.Text = "Echo ON" Then e.Handled = True ' if echo is off don't display on textbox
        If SerialPort1.IsOpen Then
            SerialPort1.Write(e.KeyChar)
        Else
            TextBox1.AppendText("No COM port open" + vbNewLine)
        End If
        Console.WriteLine("key done")
    End Sub

    ' character received from serial port
    Private Sub SerialPort1_DataReceived(ByVal sender As System.Object, ByVal e As System.IO.Ports.SerialDataReceivedEventArgs) Handles SerialPort1.DataReceived
        'TextBox1.AppendText(SerialPort1.ReadExisting())  'not thead safe use delegate
        Dim ReceivedData As String = SerialPort1.ReadExisting()
        Me.Invoke(adre, ReceivedData)     'invoke delegate
    End Sub
    ' delegate to display string received
    Private Sub DisplayData(ByVal sdata As String)
        TextBox1.AppendText(sdata)
    End Sub


    ' From is loading - scan for COM ports
    Private Sub Form1_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        Dim port As Integer = 0
        Dim COMmenuitem As ToolStripMenuItem = MenuStrip1.Items.Item(1)
        Dim ports() As String = SerialPort1.GetPortNames()
        TextBox1.AppendText("The following serial ports were found:" + vbNewLine + "  ")

        ' Display each port name to the console.
        For Each port1 As String In ports 'foreach(String port In ports)
            Try
                TextBox1.AppendText("  " + port1)
                Dim item As ToolStripMenuItem = New ToolStripMenuItem(port1) ' create a new menu item
                AddHandler item.Click, AddressOf Me.COMToolStripMenuItem_Click              ' add event handler
                COMmenuitem.DropDownItems.Add(item)                                         ' add to the list of COM ports
            Catch ex As Exception
                Console.WriteLine("serial port open fail {0}", SerialPort1.PortName)
            End Try
        Next


        Console.WriteLine(" {0} COM ports found - select from COM ports menu", PortToolStripMenuItem.DropDownItems.Count)
        TextBox1.AppendText(String.Format(vbNewLine + " {0} COM ports found - select from COM ports menu",
                                           PortToolStripMenuItem.DropDownItems.Count) + vbNewLine)
        ' if no COM ports were found display message
        If PortToolStripMenuItem.DropDownItems.Count = 0 Then
            TextBox1.AppendText(vbNewLine + "No COM ports found" + vbNewLine)
        End If

    End Sub

    ' a COM port menu item has been selected
    Private Sub COMToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs)
        Console.WriteLine("item selected {0} {1}", e.ToString(), sender.ToString())
        SerialPort1.Close()
        SerialPort1.PortName = sender.ToString()
        SerialPort1.Open()
        TextBox1.AppendText("Opened Port " + sender.ToString() + " baud rate " + SerialPort1.BaudRate.ToString() + vbNewLine)
        Text = "Port " + sender.ToString() + " baud rate " + SerialPort1.BaudRate.ToString()
    End Sub

    ' a Baud rate has been sleceted
    Private Sub ToolStripMenuItem2_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles ToolStripMenuItem2.Click, ToolStripMenuItem9.Click, ToolStripMenuItem8.Click, ToolStripMenuItem7.Click, ToolStripMenuItem6.Click, ToolStripMenuItem5.Click, ToolStripMenuItem4.Click, ToolStripMenuItem3.Click
        Console.WriteLine("item selected {0} {1}", e.ToString(), sender.ToString())
        SerialPort1.BaudRate = sender.ToString()
        TextBox1.AppendText(vbNewLine + " baud rate " + SerialPort1.BaudRate.ToString() + vbNewLine)
        Text = "Port " + SerialPort1.PortName.ToString() + " baud rate " + SerialPort1.BaudRate.ToString()
    End Sub

    ' Exit menu item has been selected
    Private Sub ExitToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles ExitToolStripMenuItem.Click
        Close()
    End Sub

    ' echo ON/OFF menu item has been selected
    Private Sub EchoOFFToolStripMenuItem_Click(sender As System.Object, e As System.EventArgs) Handles EchoOFFToolStripMenuItem.Click
        ' invert echo ON/OFF  display
        If EchoOFFToolStripMenuItem.Text = "Echo ON" Then
            EchoOFFToolStripMenuItem.Text = "Echo OFF"
        Else
            EchoOFFToolStripMenuItem.Text = "Echo ON"
        End If
    End Sub

    Private Sub Form1_FormClosing(sender As System.Object, e As System.Windows.Forms.FormClosingEventArgs) Handles MyBase.FormClosing
        SerialPort1.Close()
    End Sub
End Class

File: CsharpTerminal/CsharpTerminal/Form1.cs
using System;
using System.Windows.Forms;
using System.IO.Ports;

namespace CsharpTerminal
{
    public partial class Form1 : Form
    {
        // delegate to transfer received data to TextBox
        public delegate void AddDataDelegate(String myString);
        public AddDataDelegate myDelegate;
        
        public Form1()
        {
            InitializeComponent();
            this.myDelegate = new AddDataDelegate(AddDataMethod);
            // Get a list of serial port names.
            string[] ports = SerialPort.GetPortNames();

            textBox1.AppendText("The following serial ports were found:");

            // Display each port name to the console.
            foreach (string port in ports)
            {
                textBox1.AppendText("  " + port);
                ToolStripMenuItem item = new ToolStripMenuItem(port);   // create a new menu item
                item.Click += comPortsStripMenuItem_Click;  // add event handler
                comPortsStripMenuItem.DropDownItems.Add(item);  // add to the menu of COM ports
            }


            // scan for COM ports and add to menu item if found
          /*  for (int port = 0; port <= 20; port++)
				{
				serialPort1.PortName = "COM" + port.ToString();
				try
					{
					serialPort1.Open();			// attempt to open serial port
					serialPort1.BaudRate=115200;			
					textBox1.AppendText("Found " + serialPort1.PortName + "\n");
					Console.WriteLine("serial port open OK {0}", serialPort1.PortName);
					serialPort1.Close() ;														//open OK, close it
					ToolStripMenuItem item = new ToolStripMenuItem(serialPort1.PortName);	// create a new menu item
                    item.Click += comPortsStripMenuItem_Click;	// add event handler
                    comPortsStripMenuItem.DropDownItems.Add(item);	// add to the menu of COM ports
 					}
				catch ( Exception ex)
					{ Console.WriteLine("serial port open fail {0} ", serialPort1.PortName, ex); }
			}

        Console.WriteLine("COM ports found {0} ", comPortsStripMenuItem.DropDownItems.Count);
		textBox1.AppendText(String.Format("{0} COM ports found - select from COM ports menu\r\n\r\n", comPortsStripMenuItem.DropDownItems.Count));
        */
        // if no COM ports were found display message
        if(comPortsStripMenuItem.DropDownItems.Count == 0 )
           textBox1.AppendText("\r\n  No COM ports found\r\n");
        else textBox1.AppendText("\r\n  Found " + comPortsStripMenuItem.DropDownItems.Count + " COM ports \n");
        }

        // key on textbox pressed, read key and transmit down serial line
        private void textBox1_KeyPress(object sender, KeyPressEventArgs e)
        {        
        if(echoOFFToolStripMenuItem.Text == "Echo ON") e.Handled = true; // if echo is off don't display on textbox
        if( serialPort1.IsOpen)
            serialPort1.Write(e.KeyChar.ToString());
        else
            textBox1.AppendText("No COM port open" + Environment.NewLine);
        }

     // data received from serial port - display on textbox
     private void serialPort1_DataReceived(object sender, System.IO.Ports.SerialDataReceivedEventArgs e)
        {
            //textBox1.AppendText(serialPort1.ReadExisting());  // not thread safe
            string s = serialPort1.ReadExisting();
            textBox1.Invoke(this.myDelegate, new Object[] { s });
        }

     // display seral data on textbox
     public void AddDataMethod(String myString)
     {
         textBox1.AppendText(myString);
     }

     // COM port selected
     private void comPortsStripMenuItem_Click(object sender, EventArgs e)
     {
         try
         {
             Console.WriteLine("item selected {0} {1}", e.ToString(), sender.ToString());
             serialPort1.Close();
             serialPort1.PortName = sender.ToString();
             serialPort1.Open();
             textBox1.AppendText("Opened Port " + sender.ToString() + " baud rate " + serialPort1.BaudRate.ToString() + "\r\n");
             Text = "Port " + sender.ToString() + " baud rate " + serialPort1.BaudRate.ToString();
         }
         catch (Exception e1)
            //{ textBox1.AppendText("Failed to open Port " + sender.ToString() + "\r\n" + e1 + "\r\n"); }
            { textBox1.AppendText("Failed to open Port " + sender.ToString() + 
                    "\r\n   Is it already open in another program?? \r\n"); }
            // serialPort1.RtsEnable = true;
            // serialPort1.Handshake =  Handshake.RequestToSend;
        }

        // baud rate selected
        private void baudRateToolStripMenuItem_Click(object sender, EventArgs e)
     {
        Console.WriteLine("item selected {0} {1}", e.ToString(), sender.ToString());
        serialPort1.BaudRate = Convert.ToInt32(sender.ToString());
        textBox1.AppendText('\n' + " baud rate " + serialPort1.BaudRate.ToString() + '\n');
        Text = "Port " + serialPort1.PortName.ToString() + " baud rate " + serialPort1.BaudRate.ToString();

     }

     // echo selected
     private void echoONToolStripMenuItem_Click(object sender, EventArgs e)
     {
        // invert echo ON/OFF  display
        if(echoOFFToolStripMenuItem.Text == "Echo ON")
            echoOFFToolStripMenuItem.Text = "Echo OFF";
        else
            echoOFFToolStripMenuItem.Text = "Echo ON";
     }

     private void Form1_FormClosed(object sender, FormClosedEventArgs e)
     {
         serialPort1.Close();
     }

        // handshake mode selected
        private void handShakeNone(object sender, EventArgs e)
        {
            serialPort1.Handshake = Handshake.None;
        }

        private void handShakeRequestToSend(object sender, EventArgs e)
        {
            serialPort1.Handshake = Handshake.RequestToSend;
        }

        private void handShakeXonXoff(object sender, EventArgs e)
        {
            serialPort1.Handshake = Handshake.XOnXOff;
        }

        private void handShakeRequestXonXoff(object sender, EventArgs e)
        {
            serialPort1.Handshake = Handshake.RequestToSendXOnXOff;
        }
    }
}
